modules/agent/pkg/controllers/bundledeployment/controller.go

```diff
@@ -252,3 +279,72 @@ func readyError(status fleet.BundleDeploymentStatus) error {
 
 	return errors.New(msg)
 }
+
+func hasAutoReapply(bd *fleet.BundleDeployment, status *fleet.BundleDeploymentStatus) (ok bool, requeueAfter time.Duration, err error) {
+	requeueAfter = DefaultReapplyInterval
+	if bd.Spec.Options.Resync || bd.Spec.StagedOptions.Resync {
+		if bd.Spec.Options.ResyncPolicy == nil && bd.Spec.StagedOptions.ResyncPolicy == nil {
+			if status.LastApply == nil {
+				return false, requeueAfter, nil
+			}
+			if status.LastApply.Add(DefaultReapplyInterval).Before(time.Now()) {
+				status.LastApply = nil
```

Comment:
Seems weird to me that we are resetting `lastApply` field  during `shouldDeploy` method, as it is just checking whether you should redeploy or not. My suggestion to separate the logic so that:

1. You only set `status.LastApplied` and `status.ResyncCounter` after you deployed the bundle. This idealy should live in https://github.com/rancher/fleet/blob/72de385c179ea1c139e53cec813a54e34eb1ac64/modules/agent/pkg/controllers/bundledeployment/controller.go#L112 after the bundle is **actually** applied. There are cases where it is a no-op deploy, so we might want to update it only when it is being applied.
2. `shouldRedeploy` method checks against `lastApplied` and `ResyncCounter` and decide whether it needs to redeploy the bundle based on its own policy. It should not write anything to the status as it is just checking the bundleDeployment.

Reply:
tweaked the logic to ensure status updates are only happening in the DeployBundle and MonitorBundle.

also changed how lastApplied is handled.